src/OpenTracing/Util/GlobalTracer.cs

```diff
@@ -0,0 +1,125 @@
+
+        /// <summary>
+        /// Singleton instance.
+        /// <para/>
+        /// Since we cannot prevent people using <see cref="Instance"/> as a constant,
+        /// this guarantees that references obtained before, during or after initialization
+        /// all behave as if obtained <em>after</em> initialization once properly initialized.<br/>
+        /// As a minor additional benefit it makes it harder to circumvent the <see cref="ITracer"/> API.
+        /// </summary>
+        private static readonly GlobalTracer s_instance = new GlobalTracer();
+
+        /// <summary>
+        /// Returns the constant <see cref="GlobalTracer"/>.
+        /// <para/>
+        /// All methods are forwarded to the currently configured tracer.<br/>
+        /// Until a tracer is explicitly configured via <see cref="Register(ITracer)"/>,
+        /// the <see cref="NoopTracer"/> is used.
+        /// </summary>
+        /// <returns>The global tracer constant.</returns>
+        /// <seealso cref="Register(ITracer)"/>
+        public static ITracer Instance => s_instance;
+
+        /// <summary>
+        /// Identify whether a <see cref="ITracer"/> has previously been registered.
+        /// <para/>
+        /// This check is useful in scenarios where more than one component may be responsible
+        /// for registering a tracer.
+        /// </summary>
+        /// <returns>Whether a tracer has been registered.</returns>
+        public static bool IsRegistered()
+        {
+            return !(s_instance._tracer is NoopTracer);
+        }
+
+        /// <summary>
+        /// Register a <see cref="ITracer"/> to back the behaviour of the global tracer (<see cref="Instance"/>).
+        /// <para/>
+        /// Registration is a one-time operation, attempting to call it more often will result in a runtime exception.
+        /// <para/>
+        /// Every application intending to use the global tracer is responsible for registering it once
+        /// during its initialization.
+        /// </summary>
+        /// <param name="tracer">Tracer to use as global tracer.</param>
+        public static void Register(ITracer tracer)
+        {
+            if (tracer == null)
+                throw new ArgumentNullException(nameof(tracer), "Cannot register GlobalTracer <null>.");
+
+            if (tracer is GlobalTracer)
+                throw new ArgumentException("Attempted to register the GlobalTracer as delegate of itself.", nameof(tracer));
+
+            lock (s_lock)
+            {
+                if (tracer == s_instance._tracer)
+                    return;
+
+                if (IsRegistered())
+                    throw new InvalidOperationException("There is already a current global Tracer registered.");
+
+                s_instance._tracer = tracer;
+            }
+        }
+
+        private ITracer _tracer = NoopTracerFactory.Create();
+
+        // <inheritdoc/>
+        public IScopeManager ScopeManager => _tracer.ScopeManager;
```

Comment:
Given this isn't used in this class seems to suggest we are doing more than SRP, which I'd argue we are. We're trying to wrap a tracer while also trying to provide singleton global access.  Might I suggest something more like

```C#
public static class GlobalTracer/TracerProvider
{
  public static ITracer Tracer { get; set; } /* more details about how to implement this addressing the relevant concerns in another comment */
}
```

Reply:
The problem with this is that people could save that `TracerProvider.Tracer` variable locally and therefore never get a new tracer if someone replaces it. That's the reason why in this class, `GlobalTracer.Instance` always returns the one singleton `GlobalTracer`.

The instance types are not used because GlobalTracer is just a simple decorator/wrapper type.

Reply:
> The instance types are not used because GlobalTracer is just a simple decorator/wrapper type.

Correct, this is the expected behavior :)